<a href="https://colab.research.google.com/github/MiguelSanz2/APRENDIZAJE_AUTOMATICO_23_24/blob/main/Pr%C3%A1ctica1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

GRUPO 7:

Carlos Ricardo Adaro Cacho,

Juan Rivera Sánchez,

Miguel Sanz Almau,

**INTRODUCCIÓN**

En este trabajo vamos a realizar una Red Neuronal la cual permita diferenciar los diferentes tipos de ropa de la aplicación zalando.
Para ello, desarrollaremos un código que seguidamente lo explicaremos en base a sus resultados.

La creación de un modelo de Red de Neuronas Artificiales capaz de reconocer y clasificar imágenes de ropa es un problema común en el campo de la visión por computadora. En este escenario, utilizaremos el conjunto de datos Fashion-MNIST, el cual contiene 70000 imágenes de prendas de ropa en 10 categorías diferentes. El objetivo es desarrollar un modelo de clasificación de prendas de ropa en Keras, aplicando un conjunto de pasos que incluyen configuración, entrenamiento, evaluación y mejora.

Lo primero que deberemos hacer será importar todas las bibliotecas necesarias y asegurse de estar utilizando TensorFlow 2.0.0 o superior.
Los datos de entrenamiento y prueba se cargarán desde `keras.datasets.fashion_mnist`.


Las cuestiones a tener en cuenta incluyen la importancia de analizar si los datos tienen la forma esperada, la posible necesidad de preprocesar los datos utilizando la función `keras.layers.Flatten()`, y la configuración de varios modelos de red neuronal con diferentes hiperparámetros, como el número de neuronas en la capa, la función de activación, el optimizador y el número de épocas.


Este proyecto proporciona una oportunidad para explorar y entender la influencia de diferentes hiperparámetros en el rendimiento de una red neuronal en una tarea de clasificación de imágenes. Los resultados y análisis ayudarán a tomar decisiones más informadas en la configuración de futuros modelos de aprendizaje profundo.



1.-Configurar y entrenar los siguientes modelos de red de neuronas, todos con función de pérdida ‘categorical_crossentropy’, métrica de precisión y 5 épocas, analizando y reflexionando sobre los resultados

In [ ]:
# Importar las bibliotecas necesarias

import tensorflow as tf
from tensorflow import keras
from keras.datasets import fashion_mnist
from keras import models
from keras import layers

# Load the Fashion-MNIST dataset
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Reshape and normalize the input images
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

# Convert labels to categorical
train_labels = keras.utils.to_categorical(train_labels)
test_labels = keras.utils.to_categorical(test_labels)

# Define a function to create and train the model
def create_and_train_model(neurons, activation, optimizer):
    model = models.Sequential()
    model.add(layers.Dense(neurons, activation=activation, input_shape=(28*28,)))
    model.add(layers.Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer,
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5, batch_size=128)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print('Test loss:', test_loss)
    print('Test accuracy:', test_acc)

# Case 1: 10 neurons, relu activation, sgd optimizer
print('Modelo 1')
create_and_train_model(10, 'relu', 'sgd')

# Case 2: 10 neurons, relu activation, rmsprop optimizer
print('Modelo 2')
create_and_train_model(10, 'relu', 'rmsprop')

# Case 3: 10 neurons, sigmoid activation, sgd optimizer
print('Modelo 3')
create_and_train_model(10, 'sigmoid', 'sgd')

# Case 4: 10 neurons, sigmoid activation, rmsprop optimizer
print('Modelo 4')
create_and_train_model(10, 'sigmoid', 'rmsprop')

# Case 5: 512 neurons, relu activation, sgd optimizer
print('Modelo 5')
create_and_train_model(512, 'relu', 'sgd')

# Case 6: 512 neurons, relu activation, rmsprop optimizer
print('Modelo 6')
create_and_train_model(512, 'relu', 'rmsprop')

# Case 7: 512 neurons, sigmoid activation, sgd optimizer
print('Modelo 7')
create_and_train_model(512, 'sigmoid', 'sgd')

# Case 8: 512 neurons, sigmoid activation, rmsprop optimizer
print('Modelo 8')
create_and_train_model(512, 'sigmoid', 'rmsprop')

Modelo 1
Epoch 1/5
469/469 [==============================] - 1s 2ms/step - loss: 1.5656 - accuracy: 0.4771
Epoch 2/5
469/469 [==============================] - 1s 2ms/step - loss: 0.8887 - accuracy: 0.6841
Epoch 3/5
469/469 [==============================] - 1s 2ms/step - loss: 0.7528 - accuracy: 0.7301
Epoch 4/5
469/469 [==============================] - 1s 2ms/step - loss: 0.6803 - accuracy: 0.7657
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.6333 - accuracy: 0.7822
Test loss: 0.6332998275756836
Test accuracy: 0.7821999788284302
Modelo 2
Epoch 1/5
469/469 [==============================] - 2s 2ms/step - loss: 0.8549 - accuracy: 0.7171
Epoch 2/5
469/469 [==============================] - 1s 2ms/step - loss: 0.5229 - accuracy: 0.8199
Epoch 3/5
469/469 [==============================] - 1s 2ms/step - loss: 0.4753 - accuracy: 0.8363
Epoch 4/5
469/469 [==============================] - 1s 2ms/step - loss: 0.4522 - accuracy: 0.8436
Epoch 5/5
313/313 [========

Análisis de los resultados de entrenamiento y evaluación de los ocho modelos:

**Modelo 1**
- Número de Neuronas en la primera capa: 10
- Función de Activación: ReLU
- Optimizador: SGD
- Precisión en el conjunto de entrenamiento después de 5 épocas: 78.22%

**Modelo 2**
- Número de Neuronas en la primera capa: 10
- Función de Activación: ReLU
- Optimizador: RMSprop
- Precisión en el conjunto de entrenamiento después de 5 épocas: 82.69%

**Modelo 3**
- Número de Neuronas en la primera capa: 10
- Función de Activación: Sigmoid
- Optimizador: SGD
- Precisión en el conjunto de entrenamiento después de 5 épocas: 61.65%

**Modelo 4**
- Número de Neuronas en la primera capa: 10
- Función de Activación: Sigmoid
- Optimizador: RMSprop
- Precisión en el conjunto de entrenamiento después de 5 épocas: 82.58%

**Modelo 5**
- Número de Neuronas en la primera capa: 512
- Función de Activación: ReLU
- Optimizador: SGD
- Precisión en el conjunto de entrenamiento después de 5 épocas: 82.17%

**Modelo 6**
- Número de Neuronas en la primera capa: 512
- Función de Activación: ReLU
- Optimizador: RMSprop
- Precisión en el conjunto de entrenamiento después de 5 épocas: 87.70%

**Modelo 7**
- Número de Neuronas en la primera capa: 512
- Función de Activación: Sigmoid
- Optimizador: SGD
- Precisión en el conjunto de entrenamiento después de 5 épocas: 74.65%

**Modelo 8**
- Número de Neuronas en la primera capa: 512
- Función de Activación: Sigmoid
- Optimizador: RMSprop
- Precisión en el conjunto de entrenamiento después de 5 épocas: 84.30%

Observaciones y conclusiones:

- Los modelos 2, 4, 6 y 8, que utilizan la función de activación ReLU y el optimizador RMSprop, tienden a obtener un rendimiento más alto en el conjunto de entrenamiento como en el conjunto de prueba en comparación con los modelos que usan SGD.

- Los modelos con 512 neuronas en la primera capa superan en rendimiento a los modelos con solo 10 neuronas en la primera capa. Esto quiere decir que aumentar el número de neuronas puede mejorar el rendimiento, pero también puede requerir más recursos computacionales.

- La elección de la función de activación tiene un impacto significativo en el rendimiento. Los modelos que utilizan ReLU superan a los modelos que utilizan sigmoid en cuanto a la precisión.

- Los modelos con RMSprop como optimizador tienden a converger más rápido y a obtener mejores resultados que los modelos con SGD.



2.-Explicar la salida de la llamada model.summary() de cada uno de los 8 casos

In [ ]:
# Importar las bibliotecas necesarias

import tensorflow as tf
from tensorflow import keras
from keras.datasets import fashion_mnist
from keras import models
from keras import layers

# Cargar Fashion-MNIST dataset
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Normalizar y ajustar el tamaño de las imágenes
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

# Convertir las etiquetas a categóricas
train_labels = keras.utils.to_categorical(train_labels)
test_labels = keras.utils.to_categorical(test_labels)

# Definir una función para crear y entrenar el modelo
def create_and_train_model(neurons, activation, optimizer):
    model = models.Sequential()
    model.add(layers.Dense(neurons, activation=activation, input_shape=(28*28,)))
    model.add(layers.Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer,
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5, batch_size=128)
    model.summary()

# Caso 1: 10 neuronas, relu, sgd
create_and_train_model(10, 'relu', 'sgd')

# Caso 2: 10 neuronas, relu, rmsprop
create_and_train_model(10, 'relu', 'rmsprop')

# Caso 3: 10 neuronas, sigmoid, sgd
create_and_train_model(10, 'sigmoid', 'sgd')

# Caso 4: 10 neuronas, sigmoid, rmsprop
create_and_train_model(10, 'sigmoid', 'rmsprop')

# Caso 5: 512 neuronas, relu, sgd
create_and_train_model(512, 'relu', 'sgd')

# Caso 6: 512 neuronas, relu, rmsprop
create_and_train_model(512, 'relu', 'rmsprop')

# Caso 7: 512 neuronas, sigmoid, sgd
create_and_train_model(512, 'sigmoid', 'sgd')

# Caso 8: 512 neuronas, sigmoid, rmsprop
create_and_train_model(512, 'sigmoid', 'rmsprop')

Epoch 1/5
469/469 [==============================] - 2s 2ms/step - loss: 1.3985 - accuracy: 0.5350
Epoch 2/5
469/469 [==============================] - 1s 2ms/step - loss: 0.8516 - accuracy: 0.6898
Epoch 3/5
469/469 [==============================] - 1s 2ms/step - loss: 0.7278 - accuracy: 0.7387
Epoch 4/5
469/469 [==============================] - 1s 2ms/step - loss: 0.6611 - accuracy: 0.7699
Epoch 5/5
469/469 [==============================] - 1s 2ms/step - loss: 0.6172 - accuracy: 0.7877
Model: "sequential_104"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_208 (Dense)           (None, 10)                7850      
                                                                 
 dense_209 (Dense)           (None, 10)                110       
                                                                 
Total params: 7960 (31.09 KB)
Trainable params: 7960 (31.09 KB)
Non-trainable params:

A continuación, se muestra una explicación de la salida del método `model.summary()` para cada uno de los 8 casos, todos tienen en común que se entrenan durante 5 épocas:

**Caso 1: 10 neuronas, relu, sgd**
- El modelo tiene dos capas: una capa de entrada con 784 entradas (correspondientes a imágenes de 28x28 píxeles) y una capa de salida con 10 neuronas.
- La función de activación de la capa de entrada es 'relu'.
- La función de activación de la capa de salida es 'softmax'.
- El modelo tiene un total de 7960 parámetros (pesos y sesgos).

**Caso 2: 10 neuronas, relu, rmsprop**
- El modelo tiene la misma arquitectura que el Caso 1, pero utiliza el optimizador 'rmsprop' en lugar de 'sgd'.
- La información de la arquitectura y la cantidad de parámetros es la misma que en el Caso 1.

**Caso 3: 10 neuronas, sigmoid, sgd**
- Este modelo es idéntico al Caso 1 en términos de arquitectura y número de parámetros, pero utiliza la función de activación 'sigmoid' en lugar de 'relu'.
- La información de la arquitectura es la misma que en el Caso 1.

**Caso 4: 10 neuronas, sigmoid, rmsprop**
- Similar al Caso 3, pero con el optimizador 'rmsprop'.
- La información de la arquitectura y la cantidad de parámetros es la misma que en el Caso 3.

**Caso 5: 512 neuronas, relu, sgd**
- Este modelo tiene una arquitectura más profunda que los casos anteriores. Utiliza una capa oculta con 512 neuronas, además de la capa de entrada y salida.
- La función de activación de la capa de entrada es 'relu', al igual que la capa oculta. La capa de salida utiliza 'softmax'.
- El modelo tiene un total de 407,050 parámetros debido a la mayor cantidad de neuronas en la capa oculta.

**Caso 6: 512 neuronas, relu, rmsprop**
- Este caso es similar al Caso 5, pero utiliza el optimizador 'rmsprop'.
- La información de la arquitectura y la cantidad de parámetros es la misma que en el Caso 5.

**Caso 7: 512 neuronas, sigmoid, sgd**
- Este caso es similar al Caso 5, pero con la función de activación 'sigmoid'.
- La información de la arquitectura y la cantidad de parámetros es la misma que en el Caso 5.

**Caso 8: 512 neuronas, sigmoid, rmsprop**
- Similar al Caso 7, pero con el optimizador 'rmsprop'.
- La información de la arquitectura y la cantidad de parámetros es la misma que en el Caso 7.

En resumen, la salida de `model.summary()` muestra las configuraciones de los modelos, las funciones de activación, el número de parámetros y otros detalles importantes de cada red neuronal. Las diferencias clave entre los casos son las funciones de activación, el número de neuronas y los optimizadores utilizados. Estos aspectos pueden tener un impacto significativo en el rendimiento y la capacidad de generalización de los modelos.

3.-Analizar e interpretar los resultados del caso 2 y el 7 frente a sus originales si se
multiplica por 5 las épocas de entrenamiento (25)

In [ ]:
# Importar las bibliotecas necesarias
import tensorflow as tf
from tensorflow import keras
from keras.datasets import fashion_mnist
from keras import models
from keras import layers

# Cargar Fashion-MNIST dataset
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Normalizar y ajustar el tamaño de las imágenes
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

# Convertir las etiquetas las etiquetas a categóricas
train_labels = keras.utils.to_categorical(train_labels)
test_labels = keras.utils.to_categorical(test_labels)

# Definir una función para crear y entrenar el modelo
def create_and_train_model(neurons, activation, optimizer, epochs):
    model = models.Sequential()
    model.add(layers.Dense(neurons, activation=activation, input_shape=(28*28,)))
    model.add(layers.Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer,
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=epochs, batch_size=128)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print('Test loss:', test_loss)
    print('Test accuracy:', test_acc)

# Caso 2: 10 neuronas, relu, rmsprop (25 épocas)
create_and_train_model(10, 'relu', 'rmsprop', 25)

# Caso 7: 512 neuronas, sigmoid, sgd (25 épocas)
create_and_train_model(512, 'sigmoid', 'sgd', 25)


Epoch 1/25
469/469 [==============================] - 2s 3ms/step - loss: 0.7946 - accuracy: 0.7383
Epoch 2/25
469/469 [==============================] - 1s 3ms/step - loss: 0.5205 - accuracy: 0.8206
Epoch 3/25
469/469 [==============================] - 1s 2ms/step - loss: 0.4790 - accuracy: 0.8348
Epoch 4/25
469/469 [==============================] - 1s 2ms/step - loss: 0.4564 - accuracy: 0.8425
Epoch 5/25
469/469 [==============================] - 1s 2ms/step - loss: 0.4406 - accuracy: 0.8470
Epoch 6/25
469/469 [==============================] - 1s 3ms/step - loss: 0.4309 - accuracy: 0.8510
Epoch 7/25
469/469 [==============================] - 1s 3ms/step - loss: 0.4212 - accuracy: 0.8536
Epoch 8/25
469/469 [==============================] - 2s 4ms/step - loss: 0.4127 - accuracy: 0.8564
Epoch 9/25
469/469 [==============================] - 2s 4ms/step - loss: 0.4069 - accuracy: 0.8594
Epoch 10/25
469/469 [==============================] - 1s 3ms/step - loss: 0.4013 - accuracy: 0.8612

Estos son los resultados de los modelos 2 y 7 con solo 5 épocas de entrenamiento:

2 -->  0.8482 ||  7 --> 0.7556

Mientras que estos son los resultados de los modelos 2 y 7 con 25 épocas de entrenamiento:

2 --> 0.8743  ||  7 --> 0,8219

Como se puede observar, la precisión de estos modelos aumenta respectivamente a la precisón de los mismos modelos, pero con menos épocas de entrenamiento. Esto se debe a que al aumentar su número de entrenamientos, el algoritmo mejora y dará con mayor certeza el porcentaje del output esperado.
Además, nos damos cuenta de que el porcentaje de las 5 épocas aumenta más rápido que las de 25, ya que estas van aumentando con más cautela.


4.-Evaluar cada uno de los 8 modelos comparando el rendimiento del modelo en el conjunto de datos de prueba

**Modelo 1**


Precisión en el conjunto de prueba: 78.22%

**Modelo 2**

Número de Neuronas en la primera capa: 10
Función de Activación: ReLU
Optimizador: RMSprop
Precisión en el conjunto de entrenamiento después de 5 épocas: 82.69%
Precisión en el conjunto de prueba: 82.69%

**Modelo 3**

Número de Neuronas en la primera capa: 10
Función de Activación: Sigmoid
Optimizador: SGD
Precisión en el conjunto de entrenamiento después de 5 épocas: 61.65%
Precisión en el conjunto de prueba: 61.65%

**Modelo 4**

Número de Neuronas en la primera capa: 10
Función de Activación: Sigmoid
Optimizador: RMSprop
Precisión en el conjunto de entrenamiento después de 5 épocas: 82.58%
Precisión en el conjunto de prueba: 82.58%

**Modelo 5**

Número de Neuronas en la primera capa: 512
Función de Activación: ReLU
Optimizador: SGD
Precisión en el conjunto de entrenamiento después de 5 épocas: 82.17%
Precisión en el conjunto de prueba: 82.17%

**Modelo 6**

Número de Neuronas en la primera capa: 512
Función de Activación: ReLU
Optimizador: RMSprop
Precisión en el conjunto de entrenamiento después de 5 épocas: 87.70%
Precisión en el conjunto de prueba: 87.70%

**Modelo 7**

Número de Neuronas en la primera capa: 512
Función de Activación: Sigmoid
Optimizador: SGD
Precisión en el conjunto de entrenamiento después de 5 épocas: 74.65%
Precisión en el conjunto de prueba: 74.65%

**Modelo 8**

Número de Neuronas en la primera capa: 512
Función de Activación: Sigmoid
Optimizador: RMSprop
Precisión en el conjunto de entrenamiento después de 5 épocas: 84.30%
Precisión en el conjunto de prueba: 84.30%


5.- Usar cada uno de los 8 modelos para hacer predicciones sobre la 6ª imagen de
test (test_images[5])

In [ ]:
# Importar las bibliotecas necesarias
import tensorflow as tf
from tensorflow import keras
from keras.datasets import fashion_mnist
from keras import models
from keras import layers
import matplotlib.pyplot as plt
import numpy as np

# Cargar Fashion-MNIST dataset
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Normalizar y ajustar el tamaño de las imágenes
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

# Convertir las etiquetas a categóricas
train_labels = keras.utils.to_categorical(train_labels)
test_labels = keras.utils.to_categorical(test_labels)


model = None
# Definir una función para crear y entrenar el modelo
def create_and_train_model(neurons, activation, optimizer):
    global model
    model = models.Sequential()
    model.add(layers.Dense(neurons, activation=activation, input_shape=(28*28,)))
    model.add(layers.Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer,
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5, batch_size=128)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print('Test loss:', test_loss)
    print('Test accuracy:', test_acc)

# Caso 1: 10 neuronas, relu, sgd
print('Modelo 1')
create_and_train_model(10, 'relu', 'sgd')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 2: 10 neuronas, relu, rmsprop
print('Modelo 2')
create_and_train_model(10, 'relu', 'rmsprop')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 3: 10 neuronas, sigmoid, sgd
print('Modelo 3')
create_and_train_model(10, 'sigmoid', 'sgd')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 4: 10 neuronas, sigmoid, rmsprop
print('Modelo 4')
create_and_train_model(10, 'sigmoid', 'rmsprop')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 5: 512 neuronas, relu, sgd
print('Modelo 5')
create_and_train_model(512, 'relu', 'sgd')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 6: 512 neuronas, relu, rmsprop
print('Modelo 6')
create_and_train_model(512, 'relu', 'rmsprop')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 7: 512 neuronas, sigmoid, sgd
print('Modelo 7')
create_and_train_model(512, 'sigmoid', 'sgd')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]

# Caso 8: 512 neuronas, sigmoid, rmsprop
print('Modelo 8')
create_and_train_model(512, 'sigmoid', 'rmsprop')
# Obtener las predicciones del modelo para la 6ª imagen de prueba
predictions = model.predict(test_images)
predictions_array = predictions[5]
true_label = test_labels[5]
img = test_images[5]


Modelo 1
Epoch 1/5
469/469 [==============================] - 4s 6ms/step - loss: 1.4209 - accuracy: 0.5182
Epoch 2/5
469/469 [==============================] - 1s 3ms/step - loss: 0.8441 - accuracy: 0.7119
Epoch 3/5
469/469 [==============================] - 1s 3ms/step - loss: 0.7172 - accuracy: 0.7500
Epoch 4/5
469/469 [==============================] - 1s 3ms/step - loss: 0.6544 - accuracy: 0.7719
Epoch 5/5
313/313 [==============================] - 1s 3ms/step - loss: 0.6257 - accuracy: 0.7776
Test loss: 0.6257163286209106
Test accuracy: 0.7775999903678894
313/313 [==============================] - 1s 2ms/step
Modelo 2
Epoch 1/5
469/469 [==============================] - 3s 6ms/step - loss: 0.8354 - accuracy: 0.7244
Epoch 2/5
469/469 [==============================] - 2s 4ms/step - loss: 0.5321 - accuracy: 0.8194
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.4844 - accuracy: 0.8339
Epoch 4/5
469/469 [==============================] - 2s 4ms/step - loss

Ahora analizaremos los resultados para cada modelo:

**Modelo 1**
- Accuracy en el conjunto de prueba: 0.7776
- Loss en el conjunto de prueba: 0.6257

En este modelo, se utilizaron 10 neuronas en la capa oculta, activación ReLU y optimizador SGD. El modelo logró un accuracy en el conjunto de prueba del 77.76%. Esto significa que aproximadamente el 77.76% de las imágenes de prueba se clasificaron correctamente. El valor de pérdida en el conjunto de prueba fue 0.6257, lo que representa la magnitud de los errores en las predicciones.

**Modelo 2**
- Accuracy en el conjunto de prueba: 0.8422
- Loss en el conjunto de prueba: 0.4646

En este modelo, también se utilizaron 10 neuronas en la capa oculta, activación ReLU, pero con el optimizador RMSprop. El accuracy en el conjunto de prueba aumentó significativamente a 0.8422, lo que indica una mejora en el rendimiento en comparación con el Modelo 1. El valor de pérdida en el conjunto de prueba disminuyó a 0.4646, lo que sugiere que el modelo está haciendo predicciones más precisas.

**Modelo 3**
- Accuracy en el conjunto de prueba: 0.6305
- Loss en el conjunto de prueba: 1.2918

En este modelo, se utilizaron 10 neuronas en la capa oculta, activación sigmoid y optimizador SGD. El accuracy en el conjunto de prueba es inferior (0.6305) en comparación con los dos primeros modelos. El valor de pérdida es más alto (1.2918), lo que indica que este modelo tiene dificultades para hacer predicciones precisas.

**Modelo 4**
- Accuracy en el conjunto de prueba: 0.8230
- Loss en el conjunto de prueba: 0.5167

En este modelo, se utilizaron 10 neuronas en la capa oculta, activación sigmoid y optimizador RMSprop. El modelo muestra una mejora en el accuracy (0.8230) en comparación con el Modelo 3, y la pérdida en el conjunto de prueba disminuyó a 0.5167.

**Modelo 5**
- Accuracy en el conjunto de prueba: 0.8200
- Loss en el conjunto de prueba: 0.5280

Este modelo utilizó 512 neuronas en la capa oculta, activación ReLU y optimizador SGD. El accuracy (0.8200) es similar al del Modelo 4, pero el modelo tiene una mayor complejidad debido al mayor número de neuronas en la capa oculta.

**Modelo 6**
- Accuracy en el conjunto de prueba: 0.8751
- Loss en el conjunto de prueba: 0.3478

Este modelo también utilizó 512 neuronas en la capa oculta, activación ReLU, pero con el optimizador RMSprop. Logró un accuracy superior (0.8751) en comparación con el Modelo 5. La pérdida en el conjunto de prueba es más baja (0.3478), lo que indica predicciones más precisas.

**Modelo 7**
- Accuracy en el conjunto de prueba: 0.7513
- Loss en el conjunto de prueba: 0.7530

En este modelo, se utilizaron 512 neuronas en la capa oculta, activación sigmoid y optimizador SGD. El accuracy (0.7513) es inferior en comparación con los Modelos 5 y 6. El valor de pérdida en el conjunto de prueba es alto (0.7530), lo que sugiere que el modelo no está funcionando tan bien como otros.

**Modelo 8**
- Accuracy en el conjunto de prueba: 0.8503
- Loss en el conjunto de prueba: 0.4120

Este modelo también utilizó 512 neuronas en la capa oculta, activación sigmoid, pero con el optimizador RMSprop. Obtuvo un accuracy de 0.8503 en el conjunto de prueba, lo que es un rendimiento bastante bueno. La pérdida en el conjunto de prueba es de 0.4120, lo que indica una buena precisión en las predicciones.

En general, los resultados sugieren que los modelos con activación ReLU y optimizadores RMSprop tienden a funcionar mejor que los modelos con activación sigmoid y optimizador SGD. Los modelos más complejos con 512 neuronas parecen tener un mejor rendimiento que los modelos con solo 10 neuronas. Sin embargo, el rendimiento puede variar según la configuración específica del modelo y la naturaleza del conjunto de datos. Es importante considerar no solo el accuracy sino también la pérdida y la complejidad del modelo al evaluar su rendimiento.

6.- Utilícelo para ver la predicción, igual que el gráfico anterior, de las 10 primeras
imágenes del conjunto de test, para cada uno de los 8 casos. Reflexione y
comente las diferencias que observa.

In [ ]:
# Importar las bibliotecas necesarias
import tensorflow as tf
from tensorflow import keras
from keras.datasets import fashion_mnist
from keras import models
from keras import layers
import matplotlib.pyplot as plt
import numpy as np

# Cargar Fashion-MNIST dataset
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Normalizar y ajustar el tamaño de las imágenes
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

# Convertir las etiquetas a categóricas
train_labels = keras.utils.to_categorical(train_labels)
test_labels = keras.utils.to_categorical(test_labels)

# Definir una lista con las diferentes clases
class_names = ["T-shirt/top", "Trouser", "Pullover", "Dress", "Coat", "Sandal", "Shirt", "Sneaker", "Bag", "Ankle boot"]

#Usar las funciones dadas
def plot_image(i, predictions_array, true_label, img):
    predictions_array, true_label, img = predictions_array, int(true_label[i]), img
    img = img.reshape(28, 28) #Debimos modificar la función, ya que nos daba un error de tamaño
    plt.grid(False)
    plt.xticks([])
    plt.yticks([])
    plt.imshow(img, cmap=plt.cm.binary)
    predicted_label = np.argmax(predictions_array)
    if predicted_label == true_label:
        color = 'blue'
    else:
        color = 'red'
    plt.xlabel("{} {:2.0f}% ({})".format(class_names[predicted_label],
                                         100*np.max(predictions_array),
                                         class_names[true_label]),
                                         color=color)

def plot_value_array(a, predictions_array, true_label):
    predictions_array, true_label = predictions_array, int(true_label[0])  # Convert true_label to an integer
    plt.grid(False)
    plt.xticks(range(10))
    plt.yticks([])
    thisplot = plt.bar(range(10), predictions_array, color="#777777")
    plt.ylim([0, 1])
    predicted_label = np.argmax(predictions_array)
    thisplot[predicted_label].set_color('red')
    thisplot[true_label].set_color('green')

# Definir una función para crear y entrenar el modelo
def create_and_train_model(neurons, activation, optimizer):
    model = models.Sequential()
    model.add(layers.Dense(neurons, activation=activation, input_shape=(28*28,)))
    model.add(layers.Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer,
                loss='categorical_crossentropy',
                metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5, batch_size=128)
    model.summary()
    predictions = model.predict(test_images)
    plt.figure(figsize=(12, 4))

#Crear un bucle que imprima las 10 primeras imagenes y sus gráficos
    for j in range(10):
            plt.subplot(2, 10, j + 1)
            plot_image(j, predictions[j], test_labels[j], test_images[j])
            plt.subplot(2, 10, j + 11)
            plot_value_array(j, predictions[j], test_labels[j])
    plt.show()

# Case 1: 10 neuronas, relu , sgd
print('Modelo 1')
create_and_train_model(10, 'relu', 'sgd')

# Case 2: 10 neuronas, relu , rmsprop
print('Modelo 2')
create_and_train_model(10, 'relu', 'rmsprop')

# Case 3: 10 neuronas, sigmoid , sgd
print('Modelo 3')
create_and_train_model(10, 'sigmoid', 'sgd')

# Case 4: 10 neuronas, sigmoid , rmsprop
print('Modelo 4')
create_and_train_model(10, 'sigmoid', 'rmsprop')

# Case 5: 512 neuronas, relu , sgd
print('Modelo 5')
create_and_train_model(512, 'relu', 'sgd')

# Case 6: 512 neuronas, relu , rmsprop
print('Modelo 6')
create_and_train_model(512, 'relu', 'rmsprop')

# Case 7: 512 neuronas, sigmoid , sgd
print('Modelo 7')
create_and_train_model(512, 'sigmoid', 'sgd')

# Case 8: 512 neuronas, sigmoid , rmsprop
print('Modelo 8')
create_and_train_model(512, 'sigmoid', 'rmsprop')


Las diferencias que se pueden observar en los resultados de los diferentes modelos son principalmente en términos de precisión (accuracy) y velocidad de entrenamiento.
1. **Precisión en el conjunto de prueba (Test Accuracy)**:
    - Modelo 1: Tiene una precisión de alrededor del 77.76%.
    - Modelo 2: Obtiene una precisión de aproximadamente el 84.22%.
    - Modelo 3: Alcanza una precisión de aproximadamente el 63.05%.
    - Modelo 4: Logra una precisión de alrededor del 82.30%.
    - Modelo 5: Tiene una precisión de aproximadamente el 81.99%.
    - Modelo 6: Obtiene una precisión de alrededor del 87.51%.
    - Modelo 7: Alcanza una precisión de aproximadamente el 75.13%.
    - Modelo 8: Logra una precisión de aproximadamente el 85.03%.

2. **Velocidad de entrenamiento**:
    - Modelos 1 a 4 son más rápidos de entrenar, con tiempos de entrenamiento por época que van desde 1 a 2 segundos. Esto es probablemente porque tienen menos neuronas y, por lo tanto, menos parámetros en la red.
    - Modelos 5 a 8 son más lentos de entrenar, con tiempos de entrenamiento por época que van desde 4 a 6 segundos. Esto se debe a que tienen una capa densa con 512 neuronas, lo que resulta en más parámetros y cálculos más intensivos.

3. **Arquitectura de la red**:
    - Todos los modelos tienen una arquitectura similar con una capa de entrada y una capa de salida con activación softmax. La única diferencia radica en el número de neuronas en la capa oculta y la función de activación utilizada.

4. **Función de activación y Optimizador**:
    - Se han probado dos funciones de activación, 'relu' y 'sigmoid', en combinación con dos optimizadores, 'sgd' y 'rmsprop'.
    - Los modelos 1 y 2 utilizan la función de activación 'relu', mientras que los modelos 3 y 4 utilizan 'sigmoid'. En términos de optimizadores, los modelos 1 y 3 utilizan 'sgd', y los modelos 2 y 4 utilizan 'rmsprop'.
    - Los modelos 5 a 8 tienen una arquitectura similar a los modelos 1 a 4, pero con 512 neuronas en la capa oculta en lugar de 10.

En general, se puede observar que los modelos con 512 neuronas en la capa oculta tienden a tener una mejor precisión en el conjunto de prueba, pero también tienen un tiempo de entrenamiento más largo. Además, los modelos con activación 'relu' tienden a funcionar mejor que los modelos con activación 'sigmoid' en este conjunto de datos específico.

La elección de la arquitectura de red, la función de activación y el optimizador depende de las necesidades y restricciones específicas del problema, como la cantidad de datos disponibles y el tiempo disponible para entrenar el modelo.

7.- Mejore el modelo usando el de la configuración del caso 3, pero cambiando el
optimizador por ‘adam’ y la función de pérdida
‘sparse_categorical_crossentropy’. Buscar en internet las bases de dicho
optimizador y función de pérdida, explícalos con tus propias palabras y plantea
tus reflexiones respecto al resultado.

In [ ]:
# Importar las bibliotecas necesarias
import tensorflow as tf
from tensorflow import keras
from keras.datasets import fashion_mnist
from keras import models
from keras import layers

# Normalizar y ajustar el tamaño de las imágenes
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Convertir las etiquetas a categóricas
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

# Definir una función para crear y entrenar el modelo
def create_and_train_model(neurons, activation, optimizer):
    model = models.Sequential()
    model.add(layers.Dense(neurons, activation=activation, input_shape=(28*28,)))
    model.add(layers.Dense(10, activation='softmax'))
    model.compile(optimizer=optimizer,
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5, batch_size=128)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print('Test loss:', test_loss)
    print('Test accuracy:', test_acc)


print('Modelo con Adam  and sparse_categorical_crossentropy ')
create_and_train_model(10, 'sigmoid', 'adam')

Model with Adam optimizer and sparse_categorical_crossentropy loss
Epoch 1/5
469/469 [==============================] - 2s 3ms/step - loss: 1.4567 - accuracy: 0.6047
Epoch 2/5
469/469 [==============================] - 1s 3ms/step - loss: 0.8964 - accuracy: 0.7600
Epoch 3/5
469/469 [==============================] - 1s 3ms/step - loss: 0.7010 - accuracy: 0.8045
Epoch 4/5
469/469 [==============================] - 2s 5ms/step - loss: 0.5978 - accuracy: 0.8222
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.5470 - accuracy: 0.8209
Test loss: 0.5469821095466614
Test accuracy: 0.820900022983551


Optimizador Adaptativo "Adam":

El optimizador "Adam", que significa Adaptive Moment Estimation, es un algoritmo ampliamente empleado en el entrenamiento de redes neuronales. Su característica principal radica en su habilidad para ajustar la tasa de aprendizaje automáticamente durante el proceso de entrenamiento. Esto se logra mediante el cálculo de promedios móviles de primer y segundo orden de los gradientes de los pesos de la red.

En resumen, "Adam" es un optimizador eficaz y eficiente para la adaptación de los pesos de la red, lo que acelera la convergencia y resuelve problemas de ajuste más rápidamente en comparación con algoritmos más simples, como el descenso de gradiente estocástico (SGD).

Función de Pérdida "Sparse Categorical Crossentropy":

La función de pérdida "sparse_categorical_crossentropy" se utiliza comúnmente en tareas de clasificación con múltiples clases, en las cuales las etiquetas se representan como enteros en lugar de codificaciones "one-hot." Esta función de pérdida calcula la entropía cruzada entre las distribuciones de probabilidad previstas por el modelo y las etiquetas reales.

A diferencia de otras funciones de pérdida, que requieren representaciones "one-hot" para las etiquetas, "sparse_categorical_crossentropy" permite utilizar simples enteros para las etiquetas, lo que facilita la implementación y mejora la eficiencia.

Reflexiones sobre los Resultados:

Cuando se ajustan el optimizador y la función de pérdida en un modelo, es esencial considerar que esto puede influir en los resultados.
1. Velocidad de Convergencia: "Adam" a menudo converge más rápidamente en comparación con otros optimizadores, como "RMSprop." Esto puede resultar en tiempos de entrenamiento más cortos.

2. Estabilidad del Entrenamiento: La elección de "sparse_categorical_crossentropy" es beneficiosa en problemas de clasificación de múltiples clases, especialmente cuando las etiquetas se representan como enteros. Esto puede mejorar la estabilidad y eficiencia del entrenamiento.

3. Rendimiento Final: Los cambios en el optimizador y la función de pérdida pueden tener un impacto en el rendimiento final del modelo en el conjunto de datos de prueba. Evaluar y comparar los resultados con la configuración previa es crucial para determinar si la nueva configuración es una mejora real.
En general, la elección del optimizador y la función de pérdida depende del problema que se nos plantea. Experimentar con diferentes configuraciones es esencial para determinar cuál funciona mejor en un caso particular. Es fundamental mantener un seguimiento y evaluación continuos de los resultados para realizar ajustes y mejoras según sea necesario.

En conclusión, en este proyecto, hemos configurado y entrenado con éxito una red neuronal para clasificar prendas de vestir del conjunto de datos Fashion MNIST. Hemos logrado una precisión razonable y hemos evaluado el rendimiento del modelo en datos de prueba.

 Las reflexiones y comparaciones realizadas contribuyen a una mejor comprensión de cómo sintonizar y mejorar el rendimiento de los modelos de redes neuronales en aplicaciones prácticas.